### PR TITLE
backend_cairo: Clip drawn paths to context.clip_extents()

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
+ Fixes https://github.com/matplotlib/matplotlib/issues/4054
+ Fixes unused import.

### CODE REVIEW

File: lib/matplotlib/backends/backend_cairo.py [hunk 1]
```diff
@@ -134,6 +134,13 @@ def _fill_and_stroke (self, ctx, fill_c, alpha, alpha_overrides):
 
     @staticmethod
     def convert_path(ctx, path, transform):
+        extents = ctx.clip_extents()
+        bbox = Bbox([[extents[0], extents[1]], [extents[2], extents[3]]])
+        try:
+            path = path.clip_to_bbox (bbox)
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: nit: no space before `(` please.

File: lib/matplotlib/backends/backend_cairo.py [hunk 2]
```diff
@@ -134,6 +134,13 @@ def _fill_and_stroke (self, ctx, fill_c, alpha, alpha_overrides):
...
+        except ValueError:
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: What conditions does `ValueError` get raised on?

File: lib/matplotlib/backends/backend_cairo.py (hunk 1 above)
Comment (WeatherGod) [inline lib/matplotlib/backends/backend_cairo.py]: should probably fix the ones in the function above then?

Wow... they are all over the place here!

**Comment by MathieuDuponchelle** [comment]
There's an issue with this patch actually, looking into it

File: lib/matplotlib/backends/backend_cairo.py (hunk 2 above)
Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: Not entirely sure to be honest :)

Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: The message is ValueError need at least one array to concatenate by the way

**Comment by MathieuDuponchelle** [comment]
The thing is using clip_to_bbox isn't appropriate, as path isn't closed, it fixes my simple example because there was only one segment, which meant I didn't notice a rectangle was drawn, with multiple segments the problem is obvious. There's also an issue that the bounding box has to be "translated" to the actual values range, my patch didn't show it either because the width of the canvas == range of the values.

I'd really need an adult here to figure out how to bound that path correctly, the script I attached in the issue + this incorrect fix should make it really easy for someone more knowledgeable than I am with transformations etc. to come up with the correct patch, *wink*

File: lib/matplotlib/backends/backend_cairo.py (hunk 1 above)
Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: pep8 + commit hooks might be helpful if one doesn't mind losing some history, the autopep8 script does a decent job at fixing these issues.

Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: Except when it doesn't and breaks things.

We have pep8 tests as part of the test suite, but some file (like this one) are black-listed as known-bad.

Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: OK good to know

**Comment by MathieuDuponchelle** [comment]
Closing this pull request and opening a new one, addressing the issues I found.

**Pull request closed without merging**